packages/core/src/__tests__/provider.test.ts

```diff
@@ -0,0 +1,136 @@
+import type { Transaction } from "viem";
+import { polygonMumbai } from "viem/chains";
+import { afterEach, beforeEach, describe, it, vi } from "vitest";
+import { SmartAccountProvider } from "../provider/base.js";
+import type { UserOperationReceipt } from "../types.js";
+
+describe(
+  "Base Tests",
+  () => {
+    let retryMsDelays: number[] = [];
+
+    beforeEach(() => {
+      vi.useFakeTimers();
+      vi.spyOn(global, "setTimeout").mockImplementation(
+        // @ts-ignore: Mock implementation doesn't need to return a Timeout.
+        (callback: () => void, ms) => {
+          if (ms != null) {
+            retryMsDelays.push(ms);
+          }
+          callback();
+        }
+      );
+    });
+
+    afterEach(() => {
+      vi.restoreAllMocks();
+      retryMsDelays = [];
+    });
+
+    it("should apply only the initial delay for waitForUserOperationTransaction", async () => {
+      const providerMock = new SmartAccountProvider(
+        "ALCHEMY_RPC_URL",
+        "0xENTRYPOINT_ADDRESS",
+        polygonMumbai
+      );
+
+      const getUserOperationReceiptMock = vi
+        .spyOn(providerMock, "getUserOperationReceipt")
+        .mockImplementation(() => {
+          return Promise.resolve({
+            receipt: { transactionHash: "0xMOCK_USER_OP_RECEIPT" },
+          } as unknown as UserOperationReceipt);
+        });
+      expect(getUserOperationReceiptMock.getMockName()).toEqual(
+        "getUserOperationReceipt"
+      );
+
+      const getTransactionMock = vi
+        .spyOn(providerMock, "getTransaction")
+        .mockImplementation(() => {
+          return Promise.resolve({
+            hash: "0xMOCK_TXN_HASH",
+          } as unknown as Transaction);
+        });
+      expect(getTransactionMock.getMockName()).toEqual("getTransaction");
+
+      await providerMock.waitForUserOperationTransaction("0xTHIS_IS_A_TEST");
+      expect(retryMsDelays).toEqual([2_000]);
+      expect(getUserOperationReceiptMock).toHaveBeenCalledTimes(1);
+    });
+
+    it("should retry twice with exponential delay for waitForUserOperationTransaction", async () => {
+      const providerMock = new SmartAccountProvider(
+        "ALCHEMY_RPC_URL",
+        "0xENTRYPOINT_ADDRESS",
+        polygonMumbai
+      );
+
+      const getUserOperationReceiptMock = vi
+        .spyOn(providerMock, "getUserOperationReceipt")
+        .mockImplementationOnce(() =>
+          Promise.reject("Failed request, must retry")
+        )
+        .mockImplementation(() => {
+          return Promise.resolve({
+            receipt: { transactionHash: "0xMOCK_USER_OP_RECEIPT" },
+          } as unknown as UserOperationReceipt);
+        });
```

Comment:
we could probably break this out into a `givenGetUserOperationFailsNTimes(n)` function which we can reuse throughout the tests and do something like

```ts
const givenGetUserOperationFailsNTimes = (times: number) => {
    let retries = times;
    const mock = vi.spyOn(providerMock, "getUserOperationReceipt")
        .mockImplementation(() => {
            if (retries > 0) return Promise.reject("Failed request, must retry");

            return Promise.resolve({
                receipt: { transactionHash: "0xMOCK_USER_OP_RECEIPT" },
            } as unknown as UserOperationReceipt);    
        });
    // can do the assert here if it's needed
    return mock;
}
```

(i usually follow naming and structuring my tests based on https://martinfowler.com/bliki/GivenWhenThen.html)

Reply:
tweaked this a bit b/c logic is slightly than what I'm doing, but this a helpful framing for me going forward!